### PR TITLE
Bugfix/45: Filter out the empty or invalid keywords and only process with scraping for the valid ones

### LINKED ISSUE #45: [Bug] Empty keywords can be added to the database
## Issue

When uploading a file with an empty line break

```csv
ruby on rails
elixir phoenix

```

The empty line is treated as a valid keyword:

image

## Expected

- The CSV file is adequately validated
- An empty string must not be a valid keyword

### PR DESCRIPTION
**Pull request by nhantran3395**
#### What does this PR do?

This PR is to update the `POST /uploads` endpoint to filter out  any keywords in the uploaded file that are empty or only contain whitespaces and continue scraping for the rest valid keywords
Close #45 

#### How is it implemented?

Add logic in the `create-upload` handler that filters out empty or only contains whitespaces before processing

#### How it can be tested (manually)?

- Upload a file with empty keywords or keywords with only whitespace inside
- Verify that the scrape is only performed on the valid keywords (not empty, not containing only whitespaces)

#### Unit tests

Added tests to verify that keywords are being filtered

image

#### Screenshots

- The file being uploaded contains 1 empty,1 keyword with only whitespaces
image

- The scare is only done for valid keywords
image

### CODE REVIEW

File: apps/api/src/routes/uploads/handlers/create-upload.handler.ts [hunk 1]
```diff
@@ -22,6 +22,17 @@ export async function createNewUploadHandler(
   }
 
   const keywords = file.buffer.toString().split("\n") || [];
+
+  if (keywords.map((keyword) => keyword.trim()).includes("")) {
+    res.status(400).json({
+      ok: false,
+      message:
+        "file must not contains any empty line or line that has only whitespace",
```
Comment (olivierobert) [inline apps/api/src/routes/uploads/handlers/create-upload.handler.ts]: An alternative approach would be not to error out but only process valid keywords. In terms of UX, spaces or empty lines could be considered human errors 💡

File: apps/api/src/routes/uploads/handlers/create-upload.handler.test.ts [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    ...original,
+    scrape: jest.fn().mockImplementation(() => []),
+  };
+});
+
+async function sendUploadKeywordsRequest(path: string): Promise<any> {
+  const app = createServer();
+
+  const token =
+    "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.ZbNYWPJeFZcXiXpyRyNJGJGb2XTblHfHddmTlIFcStQ";
```
Comment (olivierobert) [inline apps/api/src/routes/uploads/handlers/create-upload.handler.test.ts:40]: Will this token even expire?

File: apps/api/src/routes/uploads/handlers/create-upload.handler.ts (hunk 1 above)
Comment (nhantran3395) [reply apps/api/src/routes/uploads/handlers/create-upload.handler.ts]: That's also works. Let me update it to process only the valid keywords and disregard invalid lines

File: apps/api/src/routes/uploads/handlers/create-upload.handler.test.ts (hunk 2 above)
Comment (nhantran3395) [reply apps/api/src/routes/uploads/handlers/create-upload.handler.test.ts:40]: Hi @olivierobert : The token will not expire. So, in this case, the test should always work.

I haven't implemented the token expiration with an `exp` claim in the JWT and token renewing flow this time.

File: apps/api/src/routes/uploads/handlers/create-upload.handler.ts (hunk 1 above)
Comment (nhantran3395) [reply apps/api/src/routes/uploads/handlers/create-upload.handler.ts]: Hi @olivierobert : I have updated the PR to filter out the invalid keywords (empty, or containing only whitespaces) and only process the valid keywords. Please take a look.

**Review by olivierobert — APPROVED** [review]

**Pull request merged**